File: common/src/main/java/guepardoapps/lucahome/common/converter/JsonDataToWirelessSwitchConverter.java
package guepardoapps.lucahome.common.converter;

import android.support.annotation.NonNull;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.UUID;

import guepardoapps.lucahome.common.classes.ILucaClass;
import guepardoapps.lucahome.common.classes.WirelessSwitch;
import guepardoapps.lucahome.common.utils.Logger;
import guepardoapps.lucahome.common.utils.StringHelper;

public final class JsonDataToWirelessSwitchConverter implements IJsonDataConverter {
    private static final String Tag = JsonDataToWirelessSwitchConverter.class.getSimpleName();
    private static final String SearchParameter = "{\"Data\":";

    private static final JsonDataToWirelessSwitchConverter Singleton = new JsonDataToWirelessSwitchConverter();

    public static JsonDataToWirelessSwitchConverter getInstance() {
        return Singleton;
    }

    private JsonDataToWirelessSwitchConverter() {
    }

    @Override
    public ArrayList<WirelessSwitch> GetList(@NonNull String[] stringArray) {
        if (StringHelper.StringsAreEqual(stringArray)) {
            return parseStringToList(stringArray[0]);
        } else {
            String usedEntry = StringHelper.SelectString(stringArray, SearchParameter);
            return parseStringToList(usedEntry);
        }
    }

    @Override
    public ArrayList<WirelessSwitch> GetList(@NonNull String jsonString) {
        return parseStringToList(jsonString);
    }

    private static ArrayList<WirelessSwitch> parseStringToList(String value) {
        if (!value.contains("Error")) {
            ArrayList<WirelessSwitch> list = new ArrayList<>();

            try {
                JSONObject jsonObject = new JSONObject(value);
                JSONArray dataArray = jsonObject.getJSONArray("Data");

                for (int index = 0; index < dataArray.length(); index++) {
                    JSONObject child = dataArray.getJSONObject(index).getJSONObject("WirelessSwitch");

                    UUID uuid = UUID.fromString(child.getString("Uuid"));
                    UUID roomUuid = UUID.fromString(child.getString("RoomUuid"));

                    String name = child.getString("Name");

                    int remoteId = child.getInt("RemoteId");
                    char keyCode = child.getString("KeyCode").charAt(0);

                    boolean action = child.getString("Action").contains("1");

                    JSONObject lastTriggerJsonData = child.getJSONObject("LastTrigger");

                    int year = lastTriggerJsonData.getInt("Year");
                    int month = lastTriggerJsonData.getInt("Month");
                    int day = lastTriggerJsonData.getInt("Day");
                    int hour = lastTriggerJsonData.getInt("Hour");
                    int minute = lastTriggerJsonData.getInt("Minute");

                    if (year == -1 || month == -1 || day == -1 || hour == -1 || minute == -1) {
                        year = 1970;
                        month = 1;
                        day = 1;
                        hour = 0;
                        minute = 0;
                    }

                    Calendar lastTriggerDateTime = Calendar.getInstance();
                    lastTriggerDateTime.set(year, month, day, hour, minute);

                    String lastTriggerUser = lastTriggerJsonData.getString("UserName");

                    WirelessSwitch newWirelessSwitch = new WirelessSwitch(uuid, roomUuid, name, remoteId, keyCode, action, lastTriggerDateTime, lastTriggerUser, true, ILucaClass.LucaServerDbAction.Null);
                    list.add(newWirelessSwitch);
                }

            } catch (JSONException jsonException) {
                Logger.getInstance().Error(Tag, jsonException.getMessage());
            }
            return list;
        }

        Logger.getInstance().Error(Tag, value + " has an error!");
        return new ArrayList<>();
    }
}

File: common/src/main/java/guepardoapps/lucahome/common/converter/wirelessswitch/JsonDataToWirelessSwitchConverter.kt
package guepardoapps.lucahome.common.converter.wirelessswitch

import guepardoapps.lucahome.common.annotations.JsonKey
import guepardoapps.lucahome.common.extensions.wirelessswitch.getJsonKey
import guepardoapps.lucahome.common.extensions.wirelessswitch.getPropertyJsonKey
import guepardoapps.lucahome.common.models.wirelessswitch.WirelessSwitch
import guepardoapps.lucahome.common.utils.Logger
import org.json.JSONArray
import org.json.JSONObject
import java.util.*

class JsonDataToWirelessSwitchConverter {
    private val tag = JsonDataToWirelessSwitchConverter::class.java.simpleName

    fun parse(jsonResponse: String): ArrayList<WirelessSwitch> {
        val list: ArrayList<WirelessSwitch> = ArrayList()

        if (jsonResponse.contains("Error")) {
            Logger.instance.error(tag, "Found error parsing: $jsonResponse")
            return list
        }

        try {
            val jsonObject = JSONObject(jsonResponse)

            val wirelessSwitch = WirelessSwitch()
            val jsonKey: JsonKey = wirelessSwitch.getJsonKey()
            val dataArray: JSONArray = jsonObject.getJSONArray(jsonKey.parent)

            for (index: Int in 0..dataArray.length()) {
                val value: JSONObject = dataArray.getJSONObject(index)
                val data: JSONObject = value.getJSONObject(WirelessSwitch::class.java.simpleName)

                val codeJsonKey = wirelessSwitch.getPropertyJsonKey(wirelessSwitch::code.name)
                val nameJsonKey = wirelessSwitch.getPropertyJsonKey(wirelessSwitch::name.name)
                val roomUuidJsonKey = wirelessSwitch.getPropertyJsonKey(wirelessSwitch::roomUuid.name)
                val uuidJsonKey = wirelessSwitch.getPropertyJsonKey(wirelessSwitch::uuid.name)
                val lastTriggerDateTimeJsonKey = wirelessSwitch.getPropertyJsonKey(wirelessSwitch::lastTriggerDateTime.name)
                val lastTriggerUserJsonKey = wirelessSwitch.getPropertyJsonKey(wirelessSwitch::lastTriggerUser.name)

                wirelessSwitch.uuid = UUID.fromString(data.getString(uuidJsonKey.key))
                wirelessSwitch.roomUuid = UUID.fromString(data.getString(roomUuidJsonKey.key))
                wirelessSwitch.name = data.getString(nameJsonKey.key)
                wirelessSwitch.code = data.getString(codeJsonKey.key)

                val lastTrigger: JSONObject = data.getJSONObject(lastTriggerDateTimeJsonKey.parent)

                val lastTriggerDateTimeLong: Long = lastTrigger.getLong(lastTriggerDateTimeJsonKey.key)
                wirelessSwitch.lastTriggerDateTime.timeInMillis = lastTriggerDateTimeLong

                wirelessSwitch.lastTriggerUser = lastTrigger.getString(lastTriggerUserJsonKey.key)

                list.add(wirelessSwitch)
            }
        } catch (exception: Exception) {
            Logger.instance.error(tag, exception)
        } finally {
            return list
        }
    }
}